File: ruby/perfect-numbers/perfect_numbers.rb
class PerfectNumber
  def self.classify(n)
    raise RuntimeError if n <= 0

    aliquot = (1..n / 2).select { |x| (n % x).zero? }.sum

    if aliquot > n
      'abundant'
    elsif aliquot < n
      'deficient'
    else
      'perfect'
    end
  end
end

File: clojure/perfect-numbers/src/perfect_numbers.clj
(ns perfect-numbers)

(defn classify [n]
  (if (neg? n)
    (throw (IllegalArgumentException. "Negative numbers not allowed.")))
  (let [aliquot-sum (->> (range 1 n)
                         (filter #(zero? (mod n %)))
                         (apply +))]
    (condp apply [aliquot-sum n]
      = :perfect
      > :abundant
      < :deficient)))
